### PR TITLE
RemoteLogger: bringing over from the network

### PR DESCRIPTION
**Pull request by designjockey**
https://github.com/adobe-community/issues/issues/8018

Most of the library is intact, some minor changes made are commented.

### CODE REVIEW

File: test/specs/util/remoteLogger.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
+define([
+  'nbd/util/deparam',
+  'util/window',
+  'util/remoteLogger'
+], function(deparam, win, logger) {
```
Comment (designjockey) [inline test/specs/util/remoteLogger.js]: Removed `helpers/ajax` dependency.

File: test/specs/util/remoteLogger.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  'use strict';
+
+  describe('util/remoteLogger', function() {
+    var message = 'My message',
+        channel = 'mychannel';
+
+    function getLogs() {
+      var last = jasmine.Ajax.requests.mostRecent();
```
Comment (designjockey) [inline test/specs/util/remoteLogger.js]: Changed call to `jasmine.Ajax.requests.mostRecent();` was `ajax.last()` before using helpers
https://github.com/behance/be.net/blob/master/test/jasmine/specs/be/remoteLogger.js#L14
https://github.com/behance/be.net/blob/master/test/jasmine/helpers/ajax.js#L29

File: util/remoteLogger.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    },
+
+    /**
+     * Sends queued logs to server
+     *
+     * @param {string} endpoint
+     *
+     * @return {Promise}
+     */
+    send: function(endpoint) {
```
Comment (designjockey) [inline util/remoteLogger.js]: Added support for `endpoint` param which was non-existent before as the `url` in xhr requests was always `/log`

File: util/remoteLogger.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+     */
+    error: function(channel, message, context) {
+      _push('ERROR', channel, message, context);
+      return this;
+    },
+
+    /**
+     * Sends queued logs to server
+     *
+     * @param {string} endpoint
```
Comment (designjockey) [inline util/remoteLogger.js]: Added param doc block

File: util/remoteLogger.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+     *
+     * @return {Promise}
+     */
+    send: function(endpoint) {
+      var logs = [],
+          endpoint = endpoint || '/log',
```
Comment (designjockey) [inline util/remoteLogger.js]: Defaulting endpoint to `/log` for back-compat

File: util/remoteLogger.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+      // Ensure no sending can conflict with timer
+      _clearInterval();
+
+      while (queue.length) {
+        data = queue.pop();
+        logs.push(JSON.stringify(data));
+      }
+
+      chain = xhr({
+        url: endpoint,
```
Comment (designjockey) [inline util/remoteLogger.js]: Changed to `endpoint` from `/log`

File: util/remoteLogger.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    /**
+     * Sends queued logs to server
+     *
+     * @param {string} endpoint
+     *
+     * @return {Promise}
+     */
+    send: function(endpoint) {
+      var logs = [],
+          url = endpoint || '/log',
```
Comment (designjockey) [inline util/remoteLogger.js]: Defaulting to `'/log'` for backwards compatibility

File: util/remoteLogger.js [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+      // Ensure no sending can conflict with timer
+      _clearInterval();
+
+      while (queue.length) {
+        data = queue.pop();
+        logs.push(JSON.stringify(data));
+      }
+
+      chain = xhr({
+        url: url,
```
Comment (designjockey) [inline util/remoteLogger.js]: Changed the value to `url` param instead of `'/log'`

File: util/window.js [hunk 9]
```diff
@@ -0,0 +1,104 @@
+define([
```
Comment (designjockey) [inline util/window.js]: Is it worth bringing over this util, especially because it is only being used for one call here https://github.com/behance/BeFF/pull/137/files#diff-d914e7bdec731f1cf801e1c94e465c77R163 which can be replaced with a private method in the remoteLogger file

File: util/remoteLogger.js [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+  api = {
+    log: function(level, channel, message, context) {
+      _push(level.toUpperCase(), channel, message, context);
+      return this;
+    },
+
+    /**
+     * Allows consumer to overload xhr
+     */
+    xhr: xhr,
```
Comment (designjockey) [inline util/remoteLogger.js]: Exposed xhr, so that the consumer can overload it with a custom method.

File: util/remoteLogger.js [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+
+      // Ensure no sending can conflict with timer
+      _clearInterval();
+
+      while (queue.length) {
+        data = queue.pop();
+        logs.push(JSON.stringify(data));
+      }
+
+      chain = this.xhr({
```
Comment (designjockey) [inline util/remoteLogger.js]: Calls the exposed `xhr` method which is defaulted to the xhr module

File: util/remoteLogger.js (hunk 10 above)
Comment (hzoo) [inline util/remoteLogger.js]: Maybe move this up as the first thing

**Comment by hzoo** [comment]
👍 , not sure about the window util

**Comment by designjockey** [comment]
@Attamusc thoughts on the Window util?

**Comment by Attamusc** [comment]
I'm fine with leaving the window/util in here. Could give us a better place to add window usages that can stubbed at test time

**Comment by designjockey** [comment]
Sounds good. Will leave as is.

File: util/remoteLogger.js (hunk 10 above)
Comment (Attamusc) [inline util/remoteLogger.js]: 👍  for prop being the first key

File: util/remoteLogger.js [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+        url: '/log',
+        type: 'POST',
+        data: {
+          logs: logs
+        }
+      });
+
+      chain.then(_setInterval, _setInterval);
```
Comment (Attamusc) [inline util/remoteLogger.js]: `chain.finally`?

File: util/remoteLogger.js (hunk 10 above)
Comment (designjockey) [inline util/remoteLogger.js]: Sounds good

File: util/remoteLogger.js (hunk 12 above)
Comment (designjockey) [inline util/remoteLogger.js]: Is there an advantage in using `.finally` I'm not really quite sure what the difference is?

Comment (Attamusc) [inline util/remoteLogger.js]: it's the exact same as what's here, but doesn't require the redundancy of setting the same function for resolve/reject.

Comment (designjockey) [inline util/remoteLogger.js]: oh so instead of `chain.then(_setInterval, _setInterval)` it would just be `chain.finally(_setInterval)`?
That would make sense.

Comment (Attamusc) [inline util/remoteLogger.js]: Ya sorry. Should have been more clear

Comment (designjockey) [inline util/remoteLogger.js]: You were clear, just wanted to make sure I understood correctly. Totally makes sense to use finally in this case.

**Comment by designjockey** [comment]
Changes made @Attamusc

File: util/remoteLogger.js (hunk 12 above)
Comment (designjockey) [inline util/remoteLogger.js]: Tried this and it won't work because the returned promise doesn't have `finally` method as it uses a different Promise interface.

**Comment by Attamusc** [comment]
👍
